File: apps/workflow-engine/src/storage/ExecutionStore.ts
import type { ExecutionRecord, ExecutionContext, NodeData } from '../engine/types.js';

/**
 * In-memory execution history storage for POC
 */
class ExecutionStoreClass {
  private executions = new Map<string, ExecutionRecord>();
  private maxRecords = 100; // Keep last 100 executions

  /**
   * Create a new execution record when workflow starts
   */
  start(
    executionId: string,
    workflowId: string,
    workflowName: string,
    mode: 'manual' | 'webhook' | 'cron'
  ): ExecutionRecord {
    const record: ExecutionRecord = {
      id: executionId,
      workflowId,
      workflowName,
      status: 'running',
      mode,
      startTime: new Date(),
      nodeData: {},
      errors: [],
    };
    this.executions.set(executionId, record);
    this.cleanup();
    return record;
  }

  /**
   * Update execution with final state
   */
  complete(context: ExecutionContext, workflowId: string, workflowName: string): ExecutionRecord {
    const record = this.executions.get(context.executionId) || {
      id: context.executionId,
      workflowId,
      workflowName,
      status: 'running' as const,
      mode: context.mode,
      startTime: context.startTime,
      nodeData: {},
      errors: [],
    };

    // Convert nodeStates Map to plain object
    const nodeData: Record<string, NodeData[]> = {};
    for (const [nodeName, data] of context.nodeStates) {
      nodeData[nodeName] = data;
    }

    record.status = context.errors.length > 0 ? 'failed' : 'success';
    record.endTime = new Date();
    record.nodeData = nodeData;
    record.errors = context.errors;

    this.executions.set(context.executionId, record);
    return record;
  }

  /**
   * Mark execution as failed
   */
  fail(executionId: string, error: string): ExecutionRecord | undefined {
    const record = this.executions.get(executionId);
    if (!record) return undefined;

    record.status = 'failed';
    record.endTime = new Date();
    record.errors.push({ nodeName: '_system', error, timestamp: new Date() });
    return record;
  }

  get(id: string): ExecutionRecord | undefined {
    return this.executions.get(id);
  }

  list(workflowId?: string): ExecutionRecord[] {
    const all = Array.from(this.executions.values());
    const filtered = workflowId ? all.filter((e) => e.workflowId === workflowId) : all;
    return filtered.sort((a, b) => b.startTime.getTime() - a.startTime.getTime());
  }

  delete(id: string): boolean {
    return this.executions.delete(id);
  }

  clear(): void {
    this.executions.clear();
  }

  private cleanup(): void {
    if (this.executions.size > this.maxRecords) {
      const sorted = Array.from(this.executions.entries()).sort(
        ([, a], [, b]) => a.startTime.getTime() - b.startTime.getTime()
      );
      const toDelete = sorted.slice(0, sorted.length - this.maxRecords);
      for (const [id] of toDelete) {
        this.executions.delete(id);
      }
    }
  }
}

export const ExecutionStore = new ExecutionStoreClass();

File: apps/workflow-engine/src/storage/execution_store.py
"""In-memory execution history storage for POC."""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from ..engine.types import ExecutionContext, ExecutionRecord


class ExecutionStore:
    """In-memory execution history storage for POC."""

    def __init__(self, max_records: int = 100) -> None:
        self._executions: dict[str, ExecutionRecord] = {}
        self._max_records = max_records

    def start(
        self,
        execution_id: str,
        workflow_id: str,
        workflow_name: str,
        mode: Literal["manual", "webhook", "cron"],
    ) -> ExecutionRecord:
        """Create a new execution record when workflow starts."""
        from ..engine.types import ExecutionRecord

        record = ExecutionRecord(
            id=execution_id,
            workflow_id=workflow_id,
            workflow_name=workflow_name,
            status="running",
            mode=mode,
            start_time=datetime.now(),
        )

        self._executions[execution_id] = record
        self._cleanup()
        return record

    def complete(
        self,
        context: ExecutionContext,
        workflow_id: str,
        workflow_name: str,
    ) -> ExecutionRecord:
        """Update execution with final state."""
        from ..engine.types import ExecutionRecord

        record = self._executions.get(context.execution_id)

        if not record:
            record = ExecutionRecord(
                id=context.execution_id,
                workflow_id=workflow_id,
                workflow_name=workflow_name,
                status="running",
                mode=context.mode,
                start_time=context.start_time,
            )

        # Update record
        record.status = "failed" if context.errors else "success"
        record.end_time = datetime.now()
        record.node_data = dict(context.node_states)
        record.errors = list(context.errors)

        self._executions[context.execution_id] = record
        return record

    def get(self, execution_id: str) -> ExecutionRecord | None:
        """Get an execution record by ID."""
        return self._executions.get(execution_id)

    def list(self, workflow_id: str | None = None) -> list[ExecutionRecord]:
        """List execution records, optionally filtered by workflow ID."""
        records = list(self._executions.values())

        if workflow_id:
            records = [r for r in records if r.workflow_id == workflow_id]

        # Sort by start time descending
        records.sort(key=lambda r: r.start_time, reverse=True)
        return records

    def delete(self, execution_id: str) -> bool:
        """Delete an execution record."""
        if execution_id in self._executions:
            del self._executions[execution_id]
            return True
        return False

    def clear(self) -> None:
        """Clear all execution records."""
        self._executions.clear()

    def _cleanup(self) -> None:
        """Remove old records if over max."""
        if len(self._executions) > self._max_records:
            # Sort by start time and remove oldest
            sorted_records = sorted(
                self._executions.items(),
                key=lambda x: x[1].start_time,
            )
            to_delete = sorted_records[: len(sorted_records) - self._max_records]
            for exec_id, _ in to_delete:
                del self._executions[exec_id]


# Singleton instance
execution_store = ExecutionStore()
